[MRG] Add extra requirements in setup.py

jjerphan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #17320

#### What does this implement/fix? Explain your changes.

Add extra requirements of the projects.

This list is based on requirements given in:

 - contributing.rst
 - build_doc.sh
 - README.rst


#### Any other comments?

I've used the list format of `extras_require` but this could be turned into the `dict` version; see [setuptools doc](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies). 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jjerphan committed:
Add extra requirements

This list is based on requirements given in:

 - contributing.rst
 - build_doc.sh
 - README.rst

jjerphan committed:
Move extras_require in extra_setuptools_args

jjerphan committed:
Add versions for extra requirements

KumarGanesha1996 commented on setup.py:
```diff
@@ -78,6 +84,23 @@
+                'joblib>=0.15.1',
+                'memory_profiler>=0.57.0',
+                'packaging>=20.4',
+                'numpydoc>=0.9.2'
```
why not using format on it and here yes :
                'numpy >= {}'.format(NUMPY_MIN_VERSION),
                 'scipy >= {}'.format(SCIPY_MIN_VERSION),
                 'matplotlib>={}'.format(MATPLOTLIB_MIN_VERSION),
                 'scikit-image>={}'.format(SCIKIT_IMAGE_MIN_VERSION),
                 'pandas>={}'.format(PANDAS_MIN_VERSION),
                 'seaborn>={}'.format(SEABORN_MIN_VERSION),
                 'pytest>={}'.format(PYTEST_MIN_VERSION),

jjerphan committed:
Add variables for extra requirements min version

jjerphan committed:
(lint) Add extra space

glemaitre commented on setup.py:
```diff
@@ -78,6 +95,22 @@
             'alldeps': (
                 'numpy >= {}'.format(NUMPY_MIN_VERSION),
                 'scipy >= {}'.format(SCIPY_MIN_VERSION),
+                'matplotlib>={}'.format(MATPLOTLIB_MIN_VERSION),
```
I think that we should use the f-string instead of the `.format`.

glemaitre commented on setup.py:
```diff
@@ -78,6 +95,22 @@
             'alldeps': (
```
I am wondering if we can have several key in the `dict`. We did that in `imbalanced-learn` and I find quite readable.

```python
extra_require={
    "tests": (
        f"pytest>={PYTEST_MIN_VERSION}",
        ...
    ),
    "docs": (
        f"matplotlib>={MATPLOTLIB_MIN_VERSION}",
        ...
    ),
}
```

jjerphan replied:
I got them changed, let me know what you think of it.

glemaitre reviewed (commented):
I think that we can organize those slightly better.

jjerphan committed:
Change syntax to use f-strings

jjerphan committed:
Add joblib and threadpoolctl to extras_require

Numpy as Scipy were listed in the extras_require even though
they are necessary requirements.

This just add the other necessary requirements in extras_require
to keep follow the logic.

jjerphan committed:
Propose dict keys for extras_require

glemaitre commented on setup.py:
```diff
@@ -76,8 +93,32 @@
+            'tests': (
+                f'pytest>={PYTEST_MIN_VERSION}',
+                f'pytest-cov>={PYTEST_COV_MIN_VERSION}',
+                f'Cython>={CYTHON_MIN_VERSION}',
```
I like the way this is organised. My only doubt is about `Cython`. Basically we need it for building scikit-learn rather than `tests` but I am not really sure how to handle that.

I would prefer a second pair of eye on this @jeremiedbb @thomasjpfan @jnothman @NicolasHug

glemaitre replied:
ping @rth

rth replied:
Yes, Cython is a build dependency that's already handled by `pyproject.toml`, so it doesn't need to appear there.

Also can we just put the version as string directly instead of using variables with templating?
```
            'tests': (
                f'pytest>=3.3.0',
                f'pytest-cov>=xxx',
```
the reason we did that for scipy and numpy was to allow dynamically change it based on the Python implementation (and because it's used in other parts of setup.py) that doesn't need we need to do it for all other packages.

rth commented on setup.py:
```diff
@@ -76,8 +93,32 @@
-                'numpy >= {}'.format(NUMPY_MIN_VERSION),
-                'scipy >= {}'.format(SCIPY_MIN_VERSION),
+                f'numpy>={NUMPY_MIN_VERSION}',
+                f'scipy>={SCIPY_MIN_VERSION}',
```
Are you sure we need this? I would have though that them being in `install_requires` they would be installed in any case?

glemaitre replied:
It is what was thinking but it seems that they were there before. I am not sure the reason.

jjerphan replied:
They were there before and I've just adapted the code for the syntax and to add `joblib` and `threadpoolctl`.
But they needn't be there.

rth commented on setup.py:
```diff
@@ -76,8 +93,32 @@
+                f'pandas>={PANDAS_MIN_VERSION}',
+                f'seaborn>={SEABORN_MIN_VERSION}',
+                f'Pillow>={PILLOW_MIN_VERSION}',
+                f'memory_profiler>={MEMORY_PROFILER_MIN_VERSION}'
```
It shouldn't be used in examples, only to build the docs (and should be optional) unless I'm missing something.

glemaitre replied:
yes it is required by sphinx-gallery. I don't remember if `sphinx-gallery` raises only a warning or an error if we set to use it in the `conf.py` but it is not installed.

@lucyleeow @lesteve could answer quickly most probably.

rth replied:
Yes, but sphinx-gallery is not necessary to run the scripts in the example/ folder, so at least it's not in the right section...

lucyleeow replied:
It raises a warning.

glemaitre replied:
> Yes, but sphinx-gallery is not necessary to run the scripts in the example/ folder, so at least it's not in the right section...

True

> It raises a warning.

So we can consider it optional and omit it.

lucyleeow replied:
I think `pillow` is also only required for doc building and not required to just run the examples.

lucyleeow replied:
> Uhm I think that we might need it to run the example with images if I
recall correctly.

I misunderstood then. I thought 'run the scripts in the example/ folder' means e.g., download the `.py` example and running it. You do need it to build the example documentation but I didn't think you need it to just run the code in the example?

jjerphan replied:
Pillow isn't used in the examples. I'll move it in the doc section.

glemaitre replied:
yes we have `externals/_pilutils.py` to read the images. I was misleading sorry.

rth commented on setup.py:
```diff
@@ -76,8 +93,32 @@
+                f'Cython>={CYTHON_MIN_VERSION}',
+                f'flake8>={FLAKE8_MIN_VERSION}',
+                f'mypy>={MYPY_MIN_VERSION}',
+                f'coverage>={COVERAGE_MIN_VERSION}'
```
It will already be pulled by `pytest-cov` and since we don't pint it, I don't see the point of adding it here.

rth commented on setup.py:
```diff
@@ -76,8 +93,32 @@
+            'docs': (
+                f'sphinx>={SPHINX_MIN_VERSION}',
+                f'sphinx-gallery>={SPHINX_GALLERY_MIN_VERSION}',
+                f'packaging>={PACKAGING_MIN_VERSION}',
```
packaging is vendored in setuptools since 7.0 do we really need it? The contributing instructions may be inaccurate.

jjerphan replied:
OK, it is mentioned in the build script for the doc:
https://github.com/scikit-learn/scikit-learn/blob/e9340e3f4b1d7c285dc02062cd594ab457a5c268/build_tools/circle/build_doc.sh#L166-L172

and it contributing guidelines
https://github.com/scikit-learn/scikit-learn/blob/e1c44ba5037e22ac2c59d104f3fd96866fc9f76d/doc/developers/contributing.rst#L585

Should I remove those comments too?

rth replied:
No actually looks like it's still needed because we need a more recent version than the one shiped in setuptools https://github.com/scikit-learn/scikit-learn/pull/15566#discussion_r344143887

glemaitre commented:
Uhm I think that we might need it to run the example with images if I
recall correctly.

On Wed, 27 May 2020 at 15:38, Lucy Liu <notifications@github.com> wrote:

> *@lucyleeow* commented on this pull request.
> ------------------------------
>
> In setup.py
> <https://github.com/scikit-learn/scikit-learn/pull/17328#discussion_r431137030>
> :
>
> > @@ -76,8 +93,32 @@
>          include_package_data=True,
>          extras_require={
>              'alldeps': (
> -                'numpy >= {}'.format(NUMPY_MIN_VERSION),
> -                'scipy >= {}'.format(SCIPY_MIN_VERSION),
> +                f'numpy>={NUMPY_MIN_VERSION}',
> +                f'scipy>={SCIPY_MIN_VERSION}',
> +                f'joblib>={JOBLIB_MIN_VERSION}',
> +                f'threadpoolctl>={THREADPOOLCTL_MIN_VERSION}'
> +            ),
> +            'examples': (
> +                f'matplotlib>={MATPLOTLIB_MIN_VERSION}',
> +                f'scikit-image>={SCIKIT_IMAGE_MIN_VERSION}',
> +                f'pandas>={PANDAS_MIN_VERSION}',
> +                f'seaborn>={SEABORN_MIN_VERSION}',
> +                f'Pillow>={PILLOW_MIN_VERSION}',
> +                f'memory_profiler>={MEMORY_PROFILER_MIN_VERSION}'
>
> I think pillow is also only required for doc building and not required to
> just run the examples.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/17328#discussion_r431137030>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABY32P35G7E37LV7J6WRZQTRTUJT3ANCNFSM4NJBHOIQ>
> .
>


-- 
Guillaume Lemaitre
Scikit-learn @ Inria Foundation
https://glemaitre.github.io/

jjerphan committed:
Remove packing from extras_require

jjerphan committed:
Remove Cython from extras_require

jjerphan committed:
Remove install dependencies

jjerphan committed:
Move packages to appropriate sections

jjerphan committed:
Remove variables for packages versions

jjerphan commented:
@glemaitre, @rth : have you had some time to have a look at the last changes?

rth reviewed (approved):
LGTM, thanks @jjerphan !

thomasjpfan commented on setup.py:
```diff
@@ -256,10 +273,10 @@ def setup_package():
-                        'scipy>={}'.format(SCIPY_MIN_VERSION),
-                        'joblib>={}'.format(JOBLIB_MIN_VERSION),
-                        'threadpoolctl>={}'.format(THREADPOOLCTL_MIN_VERSION)
+                        f'numpy>={NUMPY_MIN_VERSION}',
+                        f'scipy>={SCIPY_MIN_VERSION}',
+                        f'joblib>={JOBLIB_MIN_VERSION}',
+                        f'threadpoolctl>={THREADPOOLCTL_MIN_VERSION}'
```
When building from source, this would result in a `SyntaxError`:

```
    f'numpy>={NUMPY_MIN_VERSION}',
                                ^
SyntaxError: invalid syntax
```

rather than the nicer:

```
ERROR: Package 'scikit-learn' requires a different Python: 3.5.5
```

jjerphan replied:
I've reinstalled the source in a new virtual environment with the editable mode [as indicated in the documentation](https://scikit-learn.org/stable/developers/advanced_installation.html#building-from-source) using:
```bash
python -m venv /tmp/venv
source /tmp/venv/bin/activate
pip install wheel
pip install cython
pip install --verbose --no-build-isolation --editable .
```
but I haven't meet this problem.

How can I reproduce this problem?

rth replied:
Thomas is right that f-stings are not supported in Python <=3.5: we should revert to using `.format` so that people who mistakenly use Py3.5 get a nice error message.

jjerphan replied:
As far as I've seen, scikit-learn requires python 3.6 and newer, isn't it?

rth replied:
Yes, but if you mistakenly try install it for Python 3.5- previously one would get an error saying that you need python 3.6+ now it will just produce a syntax error which is not very nice.

jjerphan replied:
OK, make sense. I followed @glemaitre 's suggestions (see above) but I can revert back to the `format` syntax.

jjerphan committed:
Revert back to format syntax

This way python <=3.5 users will get appropriate error messages.

thomasjpfan reviewed (approved):
LGTM Thank you @jjerphan !
